### PR TITLE
The initial/selected item on popup menu should always be visible

### LINKED ISSUE #142895: [Material] showMenu() doesn't scroll initial item to visible
## Steps to reproduce

Run the following app and click to open the popup menu:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: Scaffold(body: BugReport())));
}

class BugReport extends StatefulWidget {
  const BugReport({super.key});

  @override
  State<BugReport> createState() => _BugReportState();
}

class _BugReportState extends State<BugReport> {
  static const int _length = 50;

  List<PopupMenuEntry<int>> _buildItems(BuildContext context) {
    return List<PopupMenuEntry<int>>.generate(_length, (int index) {
      return PopupMenuItem<int>(value: index, child: Text('$index'));
    });
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(50),
      child: Align(
        alignment: Alignment.bottomCenter,
        child: PopupMenuButton(
          itemBuilder: _buildItems,
          initialValue: _length - 1,
          child: const Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text('Click here to open popup menu '),
              Icon(Icons.ads_click),
            ],
          ),
        ),
      ),
    );
  }
}
```

## Expected behavior

You expect that since item 49 is initially selected, it will also be scrolled to visible.

## Actual behavior

Although item 49 is correctly selected when you open the popup menu, the list of items is not scrolled at all, which means the user can't see that it's selected without scrolling to find it themselves, leading to a confusion experience.

## Test reproduction

Once fixed, this test should pass (as of this filing, the test fails):

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('PopupMenuButton scrolls initial value to visible', (WidgetTester tester) async {
    const int length = 50;
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: Align(
            alignment: Alignment.bottomCenter,
            child: PopupMenuButton(
              itemBuilder: (BuildContext context) {
                return List<PopupMenuEntry<int>>.generate(length, (int index) {
                  return PopupMenuItem<int>(value: index, child: Text('item #$index'));
                });
              },
              popUpAnimationStyle: AnimationStyle.noAnimation,
              initialValue: length - 1,
              child: const Text('click here'),
            ),
          ),
        ),
      ),
    );
    await tester.tap(find.text('click here'));
    await tester.pump();

    // Set up finder and verify basic widget structure
    final Finder findSelectedItem = find.text('item #49');
    expect(findSelectedItem, findsOne);

    // The initially selected menu item should be positioned on screen
    final RenderBox selectedItem = tester.renderObject<RenderBox>(findSelectedItem);
    final Rect selectedItemBounds = selectedItem.localToGlobal(Offset.zero) & selectedItem.size;
    final Size windowSize = tester.view.physicalSize / tester.view.devicePixelRatio;
    expect(selectedItemBounds.bottomRight.dy, lessThanOrEqualTo(windowSize.height));
  });
}
```

### PR DESCRIPTION
**Pull request by QuncCccccc**
Fixes #142895

With the change of #143121, this PR is to add auto scroll to `PopupMenuButton` so when we open the menu, it will automatically scroll to the selected item.

https://github.com/flutter/flutter/assets/36861262/c2bc0395-0641-4e7a-a54d-57a8e62ee26f

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 1]
```diff
@@ -849,6 +854,14 @@ class _PopupMenuRoute<T> extends PopupRoute<T> {
     return super.createAnimation();
   }
 
+  void scrollTo(int selectedItemIndex) {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/popup_menu.dart:858]: I'm not sure what the distinction between this and SchedulerBinding.instance.addPostFrameCallback is. Most of our code appears to use the SchedulerBinding version.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 2]
```diff
@@ -811,6 +816,7 @@ class _PopupMenuRoute<T> extends PopupRoute<T> {
 
   final RelativeRect position;
   final List<PopupMenuEntry<T>> items;
+  final List<GlobalKey>? itemKeys;
```
Comment (tvolkert) [inline packages/flutter/lib/src/material/popup_menu.dart:819]: Are there any other users of `_PopupMenuRoute`? If not, this doesn't need to be nullable.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 3]
```diff
@@ -985,10 +1004,12 @@ Future<T?> showMenu<T>({
       semanticLabel ??= MaterialLocalizations.of(context).popupMenuLabel;
   }
 
+  final List<GlobalKey> menuItemKeys = itemKeys ?? List<GlobalKey>.generate(items.length, (int index) => GlobalKey());
```
Comment (tvolkert) [inline packages/flutter/lib/src/material/popup_menu.dart:1007]: If we create global keys here, don't we need to dispose them?

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 4]
```diff
@@ -958,6 +976,7 @@ Future<T?> showMenu<T>({
   required BuildContext context,
   required RelativeRect position,
   required List<PopupMenuEntry<T>> items,
+  List<GlobalKey>? itemKeys,
```
Comment (tvolkert) [inline packages/flutter/lib/src/material/popup_menu.dart:979]: What's the use case for exposing these in the API?

If we do expose them, please cover them in the dart docs above.

**Review by tvolkert — CHANGES_REQUESTED** [review]
Does this keep the behavior that the selected item will appear over the button when a `PopupMenuButton` is used?

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 3 above)
Comment (tvolkert) [reply packages/flutter/lib/src/material/popup_menu.dart:1007]: Weird - I thought you needed to dispose GlobalKeys, but it looks like I'm wrong - please disregard 🙂

**Comment by QuncCccccc** [comment]
> Does this keep the behavior that the selected item will appear over the button when a `PopupMenuButton` is used?

Yes, the selected item will still appear over the button but the menu position would be different. With this PR's change, the selected item will always on the top of the menu and the menu is aligned with the button. The difference is the top of the menu is always the same as the button top unless there is not enough space for menu, then the menu would be pushed up like the original way.

This demo below is shown in previous PR [comment](https://github.com/flutter/flutter/pull/143121#issuecomment-1933138597)
<details><summary>video demo</summary>

https://github.com/flutter/flutter/assets/36861262/0782ed76-0616-4892-925f-520073984e3a
</details>

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 2 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/popup_menu.dart:819]: You are right. Thanks for catching this! Updating.

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 4 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/popup_menu.dart:979]: Oh sorry, I think we don't have to expose this. I used it to do some experiments but forgot to clean it up! Updating.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 5]
```diff
@@ -573,12 +574,14 @@ class _PopupMenu<T> extends StatelessWidget {
...
     required this.semanticLabel,
     this.constraints,
+    this.itemKeys,
     required this.clipBehavior,
   });
 
   final _PopupMenuRoute<T> route;
   final String? semanticLabel;
   final BoxConstraints? constraints;
+  final List<GlobalKey>? itemKeys;
```
Comment (tvolkert) [inline packages/flutter/lib/src/material/popup_menu.dart:584]: Looks like this can be non-nullable now too?

Comment (QuncCccccc) [reply packages/flutter/lib/src/material/popup_menu.dart]: Yes, just updated. Thanks!!

**Review by tvolkert — APPROVED** [review]
LGTM - thanks!

**Pull request merged**